docs/rfcs/20220413-aot-for-all-snode.md

```diff
@@ -0,0 +1,340 @@
+
+Here's the Python AOT API to save the SNodeTree type.
+
+```py
+builder = ti.SNodeTreeBuilder()
+# ...
+tree_t = builder.build()
+
+@ti.kernel
+def foo(tr: tree_t):
+  # ...
+
+m = ti.aot.Module(arch)
+m.add_snode_tree_type(tree_t, name="vel_tree")
+m.add_kernel(foo)
+m.save('/path/to/module')
+```
+
+## C++ AOT API
+
+```cpp
+auto mod = taichi::aot::Module("/path/to/module");
+auto *tree_t = mod->get_snode_tree("vel_tree");
+taichi::Device::AllocParams alloc_params;
+alloc_params.size = tree_t->get_size();
+auto *tree_mem = device->allocate_memory(alloc_params);
+// By doing this, the kernel can verify that the passed in memory matches its
+// signature.
+auto *typed_tree_mem = taichi::make_memory_typed(tree_t, tree_mem);
+
+auto foo_kernel = mod->get_kernel("foo");
+foo_kernel->launch(/*args=*/{typed_tree_mem});
+```
+
+## Backward Compatibility
+
+We need to make sure `ti.SNodeTreeBuilder` can still support the existing usage. Right now `ti.root` is already implemented as a "field accumulator": All fields being accumulated in root get materialized into a new SNode tree upon kernel invocation.
+
+Let's start with a simple example:
+
+```py
+x = ti.field(ti.f32)
+ti.root.pointer(ti.i, 4).dense(ti.i, 8).place(x)
+
+@ti.kernel
+def foo():
+  for i in x:
+    x[i] = i * 2.0
+```
+
+The equivalent code using the new `SNodeTreeBuilder` API is shown below:
+
+```py
+b = ti.SNodeTreeBuilder()
+x = b.add_field(ti.f32, name='x')
+b.tree().pointer(ti.i, 4).dense(ti.i, 8).place(x)
+tree_t = b.build()
+
+tr = tree_t.instantiate()
+
+@ti.kernel
+def foo():
+  for i in tr.x:
+    tr.x[i] = i * 2.0
+```
+
+In order to provide backward compatibility, we need some helper utils to make the following happen:
+
+* Maps `x@old` to `tr.x@new`. In addition, the runtime will need to know which SNode tree `x@old` belongs to.
+* `x@old` returned by `ti.field()` will be a placeholder for field, until the current SNode tree of `ti.root` is built and instantiated.
+
+All these being considered, here's a possible solution.
+
+`ti.root` is just a global `SNodeTreeBuilder`.
+
+`ti.field()` returns a `FieldThunk` ([what is a *thunk*?](https://en.wikipedia.org/wiki/Thunk)).
+
+```py
+class FieldThunk:
+  def __init__(self, fid):
+    self.field_id = fid
+    self.tree = None
+
+  def bind(self, tree):
+    self.tree = tree
+
+def field(dtype, name='', shape=None, offset=None, needs_grad=False):
+  name = name or random_name()
+  f = ti.root.add_field(dtype, name)
+  ft = FieldThunk(f)
+  ti.root._field_thunks.append(ft)
+  return ft
+```
+
+Upon materializing the SNodeTree:
+
+```py
+tree_t = ti.root.build()
+tree = tree_t.instantiate()
+ti._runtime.global_snode_trees.append(tree)
+for ft in ti.root._field_thunks:
+  ft.bind(tree)
+
+# Make `ti.root` a new SNodeTreeBuilder to allow for dynamic fields
+ti.root = SNodeTreeBuilder()
+```
+
+When JIT compiling a Taichi kernel, it transforms `x@old` into `x.tree.get_field(x.field_id)`, where `x` is the `FieldThunk`.
```

Comment:
OOC what does field_id map to in our current version?

Reply:
Nothing... It will be a handle into a SNodeTree to identify a field